### PR TITLE
Fix 'use auto prop' case that could cause definite assignment errors

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/32597

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UseAutoProperty/CSharpUseAutoPropertyAnalyzer.cs [hunk 1]
```diff
@@ -95,33 +96,63 @@ SyntaxKind.RecordDeclaration or
...
+            if (!memberAccess.IsOnlyWrittenTo())
+                return;
+
+            var symbolInfo = semanticModel.GetSymbolInfo(memberAccess, cancellationToken);
+            if (symbolInfo.GetAnySymbol() is not IFieldSymbol)
+                return;
+
+            // this only matters for a struct.  They can be declared unassigned and have their fields directly written into.
+            var exprType = semanticModel.GetTypeInfo(memberAccess.Expression, cancellationToken).Type;
+            if (exprType?.TypeKind != TypeKind.Struct)
```
Comment (sharwell) [inline src/Analyzers/CSharp/Analyzers/UseAutoProperty/CSharpUseAutoPropertyAnalyzer.cs:121]: ❔ Can we avoid the call to `GetTypeInfo` by using `IFieldSymbol.DeclaringType` from the previous check?

File: src/Analyzers/CSharp/Analyzers/UseAutoProperty/CSharpUseAutoPropertyAnalyzer.cs [hunk 2]
```diff
@@ -95,33 +96,63 @@ SyntaxKind.RecordDeclaration or
                         if (addressOfExpression.Kind() == SyntaxKind.AddressOfExpression)
                             AddIneligibleFields(semanticModel, addressOfExpression.Operand, ineligibleFields, cancellationToken);
                     }
+
+                    foreach (var memberAccess in typeDeclaration.DescendantNodesAndSelf().OfType<MemberAccessExpressionSyntax>())
+                        AddIneligibleFieldsIfAccessedOffNotDefinitelyAssignedValue(semanticModel, memberAccess, ineligibleFields, cancellationToken);
```
Comment (sharwell) [inline src/Analyzers/CSharp/Analyzers/UseAutoProperty/CSharpUseAutoPropertyAnalyzer.cs:101]: I worry about the placement of this. I would expect this to be efficient if we made it an IOperation-based symbol start/stop analyzer, where we track these ineligible fields as we come across them and not try and re-walk the type later.

**Review by sharwell — APPROVED** [review]
Approved on condition that @mavasani reviews the usage of semantic model and agrees that performance will either not be worse than current performance, and/or has tracking work to convert this analyzer to a more efficient form and finds that this change will not make the future work any more difficult than it otherwise would be.

File: src/Analyzers/CSharp/Analyzers/UseAutoProperty/CSharpUseAutoPropertyAnalyzer.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/UseAutoProperty/CSharpUseAutoPropertyAnalyzer.cs:121]: yup.  that makes sense.  will fix.

File: src/Analyzers/CSharp/Analyzers/UseAutoProperty/CSharpUseAutoPropertyAnalyzer.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/UseAutoProperty/CSharpUseAutoPropertyAnalyzer.cs:101]: it's effectively equally as inefficient as our current approach for everything else in this type.  :)

**Comment by CyrusNajmabadi** [comment]
Have no issue if anyone wants to change this to be iop based, or move to symbol start/end stuff :)

**Pull request merged**